File: src/scrapers/china.ts
import axios from 'axios';
import Bluebird from 'bluebird';

import { USER_AGENT } from '../constants/userAgent';
import { getLatLng } from '../helpers/geocoder';

const CHINA_URL = 'http://www.cnemc.cn/getIndexData.do';

interface CityDataPoint {
  PM25: string;
  PM10: string;
  o3: string;
  AQI: string;
  SO2: string;
  NO2: string;
  CO: string;
  CITYNAME: string;
  airCity: {
    CITY_NAME: string;
    EN_CITY_NAME: string;
  };
}

interface Response {
  airList: CityDataPoint[];
}

const geocoderExceptions: Record<string, string> = {
  玉树州: 'Yushu, China',
};

export default async function china(): Promise<App.City[]> {
  const result = await axios.get<Response>(CHINA_URL, {
    headers: {
      'User-Agent': USER_AGENT,
      Accept: 'application/json',
    },
  });

  console.log('Scraped', result.data.airList.length, 'for China');

  return Bluebird.map(
    result.data.airList,
    async (cityDataPoint): Promise<App.City> => {
      let location: App.City['location'] = {
        lat: 0,
        lng: 0,
      };

      try {
        location = await getLatLng(
          geocoderExceptions[cityDataPoint.CITYNAME] ?? cityDataPoint.CITYNAME
        );
      } catch (e) {
        console.error(e);
        console.log('Using dummy location for', cityDataPoint.CITYNAME);
      }

      return {
        name: cityDataPoint.CITYNAME,
        region: 'China',
        data: parseInt(cityDataPoint.PM25, 10) || 0,
        location,
      };
    },
    { concurrency: 2 }
  );
}

File: internal/sources/china.go
package sources

import (
	"context"

	"github.com/undertideco/nebulo-scraper/internal/httpclient"
)

const chinaURL = "http://www.cnemc.cn/getIndexData.do"

var chinaGeocoderExceptions = map[string]string{
	"玉树州": "Yushu, China",
}

func scrapeChina(ctx context.Context, r *Runner) ([]City, error) {
	var resp chinaResponse
	if err := httpclient.GetJSON(ctx, r.Client, chinaURL, &resp); err != nil {
		return nil, err
	}

	r.Logger.Printf("Scraped %d for China", len(resp.AirList))

	geocoder, err := r.RequireGeocoder()
	if err != nil {
		return nil, err
	}

	return mapConcurrent(ctx, resp.AirList, 2, func(ctx context.Context, point chinaCityDataPoint) (City, error) {
		address := point.CityName
		if replacement, ok := chinaGeocoderExceptions[address]; ok {
			address = replacement
		}

		location, err := geocoder.LatLng(ctx, address)
		if err != nil {
			r.Logger.Printf("Using dummy location for %s: %v", point.CityName, err)
			location = Location{}
		}

		return City{
			Name:     point.CityName,
			Region:   "China",
			Data:     atoiDefault(point.PM25, 0),
			Location: location,
		}, nil
	})
}

type chinaResponse struct {
	AirList []chinaCityDataPoint `json:"airList"`
}

type chinaCityDataPoint struct {
	PM25     string `json:"PM25"`
	CityName string `json:"CITYNAME"`
	AirCity  struct {
		CityName   string `json:"CITY_NAME"`
		EnCityName string `json:"EN_CITY_NAME"`
	} `json:"airCity"`
}
